### PR TITLE
REVSDL-2388 ActivateServiceRequest implementation

### PR DESCRIPTION
**Pull request by KKarlash**
@bonewell 
@DKlimenko 
@VladSemenyuk 
@giangtqh 
@thinhts

### CODE REVIEW

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+#include <stdint.h>
+#include <string>
+#include <map>
+
+#include "utils/macro.h"
+#include "json/json.h"
+#include "functional_module/function_ids.h"
+
+namespace can_cooperation {
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: vr_cooperation :)

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+ * @brief MessageHelper class
+ **/
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: **/ -> */

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+class MessageHelper {
+ public:
+  /**
+   * @brief Returns unique correlation ID for next VR request
+   *
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: I don't like emty line ;)

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+ * @param value Value with json
+ * @param key string with key name
+ *
+ * @return true if key exist
+ */
+bool IsMember(const Json::Value& value, const std::string& key);
+
+}  // namespace can_cooperation
+
+#endif /* SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_MESSAGE_HELPER_H_ */
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: Incorrect type of comment for this case. Change to `//`

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h (hunk 1 above)
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: same in header

File: src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+const char kJsonrpc[] = "jsonrpc";
+const char kMethod[] = "method";
+const char kError[] = "error";
+const char kMessage[] = "message";
+const char kData[] = "data";
+const char kAppId[] = "appID";
+const char kCode[] = "code";
+}  // namespace json_keys
+
+}
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h]: // namespace <name>

File: src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+#endif /* SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_MODULE_CONSTANTS_H_ */
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h]: Incorrect type of comment for this case. Change to `//`

File: src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h (hunk 5 above)
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h]: there is inner namespace:

namespace vr_cooperation {

namespace json_keys {
...
} // namespace json_keys

}

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+ * @brief MessageHelper class
+ **/
+class MessageHelper {
+ public:
+  /**
+   * @brief Returns unique correlation ID for next VR request
+   *
+   * @return Unique correlation ID
+   */
+  static uint32_t GetNextVRCorrelationID();
```
Comment (DKlimenko) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: Could you please add empty line and metjod description

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 8]
```diff
@@ -57,18 +77,74 @@ void BaseCommandRequest::Init() {
...
+  msg_params[kSuccess] = success;
+  msg_params[kResultCode] = result_code;
+  if (!info.empty()) {
+    msg_params[kInfo] = info;
+  }
+  Json::FastWriter writer;
+  std::string params = writer.write(msg_params);
+  message_->set_json_message(params);
+  message_->set_protocol_version(
+        application_manager::ProtocolVersion::kHMI);
```
Comment (bonewell) [inline src/components/vr_cooperation/src/base_command_request.cc]: I think you need to set correlation_id too.

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 9]
```diff
@@ -57,18 +77,74 @@ void BaseCommandRequest::Init() {
...
 
-void BaseCommandRequest::SendResponse() {
+void BaseCommandRequest::SendResponse(bool success,
+    const char* result_code,
+    const std::string& info,
+    bool is_mob_response/* = false*/) {
   LOG4CXX_AUTO_TRACE(logger_);
+  Json::Value msg_params;
+
+
```
Comment (DKlimenko) [inline src/components/vr_cooperation/src/base_command_request.cc]: Remove empty line

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 10]
```diff
@@ -57,18 +77,74 @@ void BaseCommandRequest::Init() {
...
+  Json::Value msg;
+
+  if (is_hmi_request) {
+    msg[kId] = service_->GetNextCorrelationID();
+  }
+
+  EventDispatcher<application_manager::MessagePtr, std::string>::instance()->
+  add_observer(function_id, msg[kId].asInt(), this);
+
+  msg[kJsonrpc] = "2.0";
```
Comment (DKlimenko) [inline src/components/vr_cooperation/src/base_command_request.cc]: May be it is better to create const for 2.0

File: src/components/vr_cooperation/src/message_helper.cc [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  if (!value.isObject()) {
+    return false;
+  }
+
+  return value.isMember(key);
+}
+
+}  // namespace can_cooperation
+
+
```
Comment (DKlimenko) [inline src/components/vr_cooperation/src/message_helper.cc]: Pls remove empty lines

**Comment by DKlimenko** [comment]
Reviewed

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 12]
```diff
@@ -30,18 +30,38 @@
...
+using json_keys::kCode;
+using json_keys::kError;
+using json_keys::kMessage;
+
 CREATE_LOGGERPTR_GLOBAL(logger_, "VRCooperation")
 
 BaseCommandRequest::BaseCommandRequest(
   const application_manager::MessagePtr& message)
   : message_(message) {
+  service_ = VRModule::instance()->service();
```
Comment (thinhts) [inline src/components/vr_cooperation/src/base_command_request.cc]: VRModule is not singleton
you should pass VRModule object to this constuctor

File: src/components/vr_cooperation/src/base_command_request.cc (hunk 9 above)
Comment (thinhts) [inline src/components/vr_cooperation/src/base_command_request.cc]: remove one empty line

File: src/components/vr_cooperation/src/message_helper.cc [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <string>
+#include "vr_cooperation/message_helper.h"
+// #include "vr_cooperation/vr_module_constants.h"
+
+namespace can_cooperation {
```
Comment (thinhts) [inline src/components/vr_cooperation/src/message_helper.cc]: vr_cooperation

File: src/components/vr_cooperation/src/base_command_request.cc (hunk 12 above)
Comment (KKarlash) [inline src/components/vr_cooperation/src/base_command_request.cc]: class VRModule : public functional_modules::GenericModule,
  public utils::Singleton<VRModule> {
...

File: src/components/vr_cooperation/src/message_helper.cc (hunk 13 above)
Comment (bonewell) [inline src/components/vr_cooperation/src/message_helper.cc]: vr_cooperation

File: src/components/vr_cooperation/src/message_helper.cc [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+bool IsMember(const Json::Value& value, const std::string& key) {
+  if (!value.isObject()) {
+    return false;
+  }
+
+  return value.isMember(key);
+}
+
+}  // namespace can_cooperation
```
Comment (bonewell) [inline src/components/vr_cooperation/src/message_helper.cc]: vr_cooperation

File: src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h (hunk 5 above)
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h]: I mean every close brace for namespace should be commented.

Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h]: I got you

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h [hunk 15]
```diff
@@ -94,4 +101,4 @@ bool IsMember(const Json::Value& value, const std::string& key);
 
 }  // namespace can_cooperation
 
-#endif /* SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_MESSAGE_HELPER_H_ */
+#endif // SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_MESSAGE_HELPER_H_
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: before // shoud be 2 spaces. I see you don't run cpplint.py ;)

**Comment by bonewell** [comment]
Reviewed.

**Comment by DKlimenko** [comment]
Reviewed

File: src/components/vr_cooperation/src/message_helper.cc [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+  if (!value.isObject()) {
+    return false;
+  }
+
+  return value.isMember(key);
+}
+
+}  // namespace vr_cooperation
+
+
```
Comment (thinhts) [inline src/components/vr_cooperation/src/message_helper.cc]: empty line

File: src/components/vr_cooperation/src/base_command_request.cc (hunk 12 above)
Comment (thinhts) [inline src/components/vr_cooperation/src/base_command_request.cc]: after implement VRModule class functionality, we will remove singleton.
Giang changed it in https://github.com/CustomSDL/sdl_core/pull/175

**Comment by thinhts** [comment]
Reviewed

**Comment by giangtqh** [comment]
reviewed

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h [hunk 17]
```diff
@@ -33,8 +33,11 @@
 #ifndef SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_BASE_COMMAND_REQUEST_H_
 #define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_BASE_COMMAND_REQUEST_H_
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h]: `#include <string>

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+ * @brief Check for existence of specified key in Json::Value
+ *
+ * @param value Value with json
+ * @param key string with key name
+ *
+ * @return true if key exist
+ */
+bool IsMember(const Json::Value& value, const std::string& key);
+
+}  // namespace can_cooperation
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: vr_cooperation

File: src/components/vr_cooperation/include/vr_cooperation/message_helper.h [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+#endif // SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_MESSAGE_HELPER_H_
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/message_helper.h]: 2 spaces before `//`

File: src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+const char kError[] = "error";
+const char kMessage[] = "message";
+const char kData[] = "data";
+const char kAppId[] = "appID";
+const char kCode[] = "code";
+}  // namespace json_keys
+
+}  // namespace vr_cooperation
+
+#endif // SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_VR_MODULE_CONSTANTS_H_
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/vr_module_constants.h]: 2 spaces before `//`

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 21]
```diff
@@ -54,6 +54,7 @@ using json_keys::kResult;
 using json_keys::kCode;
 using json_keys::kError;
 using json_keys::kMessage;
+using json_keys::kAppId;
```
Comment (KKarlash) [inline src/components/vr_cooperation/src/base_command_request.cc]: will be reverted

File: src/components/vr_cooperation/include/vr_cooperation/commands/activate_service_request.h [hunk 22]
```diff
@@ -34,6 +34,7 @@
 #define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_ACTIVATE_SERVICE_REQUEST_H_
 
 #include "vr_cooperation/commands/base_command_request.h"
+#include "vr_cooperation/vr_module.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/activate_service_request.h]: I think forward declaration will be better.

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h [hunk 23]
```diff
@@ -40,6 +40,7 @@
 
 #include "vr_cooperation/commands/command.h"
 #include "vr_cooperation/event_engine/event_observer.h"
+#include "vr_cooperation/vr_module.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h]: I think forward declaration will be better.

File: src/components/vr_cooperation/include/vr_cooperation/commands/register_service_request.h [hunk 24]
```diff
@@ -34,6 +34,7 @@
 #define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_REGISTER_SERVICE_REQUEST_H_
 
 #include "vr_cooperation/commands/base_command_request.h"
+#include "vr_cooperation/vr_module.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/register_service_request.h]: I think forward declaration will be better.

File: src/components/vr_cooperation/include/vr_cooperation/commands/unregister_service_request.h [hunk 25]
```diff
@@ -34,6 +34,7 @@
 #define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_UNREGISTER_SERVICE_REQUEST_H_
 
 #include "vr_cooperation/commands/base_command_request.h"
+#include "vr_cooperation/vr_module.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/unregister_service_request.h]: I think forward declaration will be better.

File: src/components/vr_cooperation/include/vr_cooperation/mobile_command_factory.h [hunk 26]
```diff
@@ -33,9 +33,10 @@
 #ifndef SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_MOBILE_COMMAND_FACTORY_H_
 #define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_MOBILE_COMMAND_FACTORY_H_
 
-#include "vr_cooperation/commands/command.h"
 #include "application_manager/message.h"
 #include "utils/macro.h"
+#include "vr_cooperation/commands/command.h"
+#include "vr_cooperation/vr_module.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/mobile_command_factory.h]: I think forward declaration will be better.

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h [hunk 27]
```diff
@@ -79,26 +88,85 @@ class BaseCommandRequest : public Command {
 
  protected:
   /**
-   * @brief send response to mobile
+   * @brief Converts Mobile string result code to HMI code
+   * @param hmi_code HMI result code
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h]: mob_code

**Comment by bonewell** [comment]
Reviewed.

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h [hunk 28]
```diff
@@ -79,26 +88,85 @@ class BaseCommandRequest : public Command {
...
+  /**
+   * @brief Parse result code from response
+   *
+   * @param value message in response from Mobile
+   * @param result_code Outgoing param with result code
+   * @param info Outgoing param with additional human readable info regarding the result(may be empty)
+   * @return true if it is success response? otherwise false
+   */
+  bool ParseMobileResultCode(const Json::Value& value,
+                       int&  result_code,
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_command_request.h]: align

**Comment by giangtqh** [comment]
reviewed

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 29]
```diff
@@ -57,18 +81,185 @@ void BaseCommandRequest::Init() {
 
 void BaseCommandRequest::Run() {
   LOG4CXX_AUTO_TRACE(logger_);
+  Execute();  // run child's logic
 }
 
 void BaseCommandRequest::on_event() {
```
Comment (giangtqh) [inline src/components/vr_cooperation/src/base_command_request.cc]: on_event() should call OnEvent() of child's logic

File: src/components/vr_cooperation/src/base_command_request.cc [hunk 30]
```diff
@@ -57,18 +81,185 @@ void BaseCommandRequest::Init() {
...
   LOG4CXX_AUTO_TRACE(logger_);
 }
 
-void BaseCommandRequest::SendResponse() {
+const hmi_apis::Common_Result::eType BaseCommandRequest::GetHMIResultCode(
+  std::string& mob_code) const {
```
Comment (bonewell) [inline src/components/vr_cooperation/src/base_command_request.cc]: mobile_code is better, isn't it?

**Comment by bonewell** [comment]
Reviewed.

**Comment by bonewell** [comment]
Reviewed.

**Pull request merged**